synapse/handlers/space_summary.py

```diff
@@ -76,67 +78,154 @@ async def get_space_summary(
+            )
 
-            room_entry = await self._build_room_entry(room_id)
-            rooms_result.append(room_entry)
+            rooms_result.extend(rooms)
+            events_result.extend(events)
 
-            # look for child rooms/spaces.
-            child_events = await self._get_child_events(room_id)
+            # add any children that we haven't already processed to the queue
+            for edge_event in events:
+                if edge_event["state_key"] not in processed_rooms:
+                    room_queue.append(edge_event["state_key"])
 
-            if suggested_only:
-                # we only care about suggested children
-                child_events = filter(_is_suggested_child_event, child_events)
+        return {"rooms": rooms_result, "events": events_result}
 
-            # The client-specified max_rooms_per_space limit doesn't apply to the
-            # room_id specified in the request, so we ignore it if this is the
-            # first room we are processing. Otherwise, apply any client-specified
-            # limit, capping to our built-in limit.
-            if max_rooms_per_space is not None and len(processed_rooms) > 1:
-                max_rooms = min(MAX_ROOMS_PER_SPACE, max_rooms_per_space)
-            else:
-                max_rooms = MAX_ROOMS_PER_SPACE
-
-            for edge_event in itertools.islice(child_events, max_rooms):
-                edge_room_id = edge_event.state_key
-
-                events_result.append(
-                    await self._event_serializer.serialize_event(
-                        edge_event,
-                        time_now=now,
-                        event_format=format_event_for_client_v2,
-                    )
+    async def _summarize_local_room(
+        self,
+        requester: Optional[str],
+        room_id: str,
+        suggested_only: bool,
+        max_children: Optional[int],
+    ) -> Tuple[Sequence[JsonDict], Sequence[JsonDict]]:
+        if not await self._is_room_accessible(room_id, requester):
+            return (), ()
+
+        room_entry = await self._build_room_entry(room_id)
+
+        # look for child rooms/spaces.
+        child_events = await self._get_child_events(room_id)
+
+        if suggested_only:
+            # we only care about suggested children
+            child_events = filter(_is_suggested_child_event, child_events)
+
+        if max_children is None or max_children > MAX_ROOMS_PER_SPACE:
+            max_children = MAX_ROOMS_PER_SPACE
+
+        now = self._clock.time_msec()
+        events_result = []  # type: List[JsonDict]
+        for edge_event in itertools.islice(child_events, max_children):
+            events_result.append(
+                await self._event_serializer.serialize_event(
+                    edge_event,
+                    time_now=now,
+                    event_format=format_event_for_client_v2,
                 )
+            )
+        return (room_entry,), events_result
```

Comment:
Why wrap `room_entry` in a tuple first? Will this eventually return multiple rooms or something?

Reply:
yeahh, it's mostly for symmetry with a `_summarize_remote_room` method in #9653, which can return multiple rooms.

It's worth noting that this method can also return zero rooms. I fiddled with making the whole result Optional, but it didn't really simplify anything.